### PR TITLE
Add docs for improved error reporting for IndexedDB large value read failures

### PR DESCRIPTION
**Pull request by chrisdavidmills**
### Description

Chrome has updated its error reporting for IndexedDB large value read failures with transient causes such as low memory and unrecoverable causes such as source blob files being deleted.

Chrome 130 updates the `DOMException` types to be more appropriate, and Chrome 132 further updates them and improves the associated error messages. The data source is https://chromestatus.com/feature/5140210640486400.

In this PR, I updated the `IDBRequest.error` types to hopefully make them easier to parse, and described and explained the large value read failure errors.

### CODE REVIEW

File: files/en-us/web/api/idbrequest/error/index.md [hunk 1]
```diff
@@ -14,22 +14,52 @@ request.
 
 ## Value
 
-A {{domxref("DOMException")}} or `null` if there is no error. The following error names are returned
-in the exception object:
+A {{domxref("DOMException")}} or `null` if there is no error. The following error names can be returned
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:17]: The phrasing "error names can be returned in the exception object" read a bit weird to me. How about "The exception object will have one of the following names, depending on what caused the error." ?

File: files/en-us/web/api/idbrequest/error/index.md [hunk 2]
```diff
@@ -14,22 +14,52 @@ request.
...
+in the exception object, depending on what caused the error.
+
+### Transaction errors
 
 - `AbortError`
-  - : If you abort the transaction, then all requests still in progress receive this error.
+  - : Thrown if you abort the transaction, then all requests still in progress receive this error.
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:23]: The asynchronous errors like this are not "thrown" in the sense that you can't try/catch them from script. So I would avoid that phrasing here and below.

From a script perspective, an error event is fired at the IDBRequest; if the developer has assigned an event handler, then the handler can inspect the request's `error` property and specifically `error.name` to see what the cause was.

File: files/en-us/web/api/idbrequest/error/index.md [hunk 3]
```diff
@@ -14,22 +14,52 @@ request.
...
+- `ReadOnlyError`
+  - : Thrown if the mutating operation was attempted in a read-only transaction.
+- `SyntaxError`
+  - : Thrown in the `keyPath` argument contains an invalid key path.
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:27]: Typo: "in" should be "if".

But more importantly, `SyntaxError` is thrown synchronously when certain methods are called. It would never appear as a request's error. So this is not useful for this page.

File: files/en-us/web/api/idbrequest/error/index.md [hunk 4]
```diff
@@ -14,22 +14,52 @@ request.
...
+- `TransactionInactiveError`
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:28]: `TransactionInactiveError` is thrown synchronously when certain methods are called. It would never appear as a request's error. So this is not useful for this page.

File: files/en-us/web/api/idbrequest/error/index.md [hunk 5]
```diff
@@ -14,22 +14,52 @@ request.
...
-in the exception object:
+A {{domxref("DOMException")}} or `null` if there is no error. The following error names can be returned
+in the exception object, depending on what caused the error.
+
+### Transaction errors
 
 - `AbortError`
-  - : If you abort the transaction, then all requests still in progress receive this error.
+  - : Thrown if you abort the transaction, then all requests still in progress receive this error.
+- `ReadOnlyError`
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:24]: `ReadOnlyError` is thrown synchronously when certain methods are called. It would never appear as a request's error. So this is not useful for this page.

File: files/en-us/web/api/idbrequest/error/index.md [hunk 6]
```diff
@@ -14,22 +14,52 @@ request.
...
+
+### Data integrity errors
+
 - `ConstraintError`
-  - : If you insert data that doesn't conform to a constraint.
-    It's an exception type for creating stores and indexes.
-    You get this error, for example, if you try to add a new key that already exists in the record.
+  - : Thrown if you insert data that doesn't conform to a constraint when creating stores and indexes.
+    For example, you will get this error if you try to add a new key that already exists in the record.
+- `DataError`
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:36]: This description for `DataError` ("data provided to an operation does not meet requirements") applies only when it is thrown synchronously when certain methods are called. It would never appear as a request's error. So this is not useful for this page.

File: files/en-us/web/api/idbrequest/error/index.md [hunk 7]
```diff
@@ -14,22 +14,52 @@ request.
...
+  - : Thrown if data provided to an operation does not meet requirements.
+- `DataCloneError`
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:38]: `DataCloneError` is thrown synchronously when certain methods are called. It would never appear as a request's error. So this is not useful for this page.

File: files/en-us/web/api/idbrequest/error/index.md [hunk 8]
```diff
@@ -14,22 +14,52 @@ request.
...
+  - : Thrown if the data being stored could not be cloned by the internal structured cloning algorithm.
+- `InvalidAccessError`
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:40]: `InvalidAccessError` is thrown synchronously when certain methods are called. It would never appear as a request's error. So this is not useful for this page.

File: files/en-us/web/api/idbrequest/error/index.md [hunk 9]
```diff
@@ -14,22 +14,52 @@ request.
...
+  - : Thrown if an invalid operation was performed on an object.
+- `InvalidStateError`
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:42]: `InvalidStateError` is thrown synchronously when certain methods are called. It would never appear as a request's error. So this is not useful for this page.

File: files/en-us/web/api/idbrequest/error/index.md [hunk 10]
```diff
@@ -14,22 +14,52 @@ request.
...
+  - : Thrown if an operation was called on an object on which it is not allowed or at a time when it is not allowed, or if a request is made on a source object that has been deleted or removed. For example, if a result is accessed before the corresponding request is completed.
 - `QuotaExceededError`
-  - : If you run out of disk quota and the user declined to grant you more space.
-- `UnknownError`
-  - : If the operation failed for reasons unrelated to the database itself.
-    A failure due to disk IO errors is such an example.
+  - : Thrown if you run out of disk quota and the user declined to grant the application more space.
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:45]: Some browsers prompt the user; others use heuristics. So this could be softened a bit.

File: files/en-us/web/api/idbrequest/error/index.md [hunk 11]
```diff
@@ -14,22 +14,52 @@ request.
...
 - `VersionError`
-  - : If you try to open a database with a version lower than the one it already has.
+  - : Thrown if you try to open a database with a version lower than the one it already has.
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:47]: Per above, not "thrown" but returned via the request's error property.

File: files/en-us/web/api/idbrequest/error/index.md [hunk 12]
```diff
@@ -14,22 +14,52 @@ request.
...
+Errors in this caregory are as follows. More recent browser versions have changed the error types thrown and improved the error messages to help developers distinguish between transient and unrecoverable cases.
+
+- `NotFoundError` or `NotReadableError`
+  - : Thrown for unrecoverable errors.
+- `DataError` or `UnknownError`
+  - : Thrown for transient errors, including general disk IO errors.
+
+### Related errors
 
 In addition to the error codes sent to the {{ domxref("IDBRequest") }} object,
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:64]: Not new, but this doesn't make much sense in this context.

"In addition to the error codes sent to the IDBRequest object..." - so far so good, that's what this page is about

"... asynchronous operations can also ..." - that doesn't make sense. This whole page is about how asynchronous operations that occur when the request is being executed, so this is like saying "In addition to A, there's also A". 

Finally "...raise exceptions." - as noted, async operations can't raise exceptions, they fire error events.

File: files/en-us/web/api/idbrequest/error/index.md [hunk 13]
```diff
@@ -14,22 +14,52 @@ request.
...
-  - : If the operation failed for reasons unrelated to the database itself.
-    A failure due to disk IO errors is such an example.
+  - : Thrown if you run out of disk quota and the user declined to grant the application more space.
 - `VersionError`
-  - : If you try to open a database with a version lower than the one it already has.
+  - : Thrown if you try to open a database with a version lower than the one it already has.
+
+### Large value read failure errors
+
+Large IndexedDB values are not stored directly in the underlying database; instead, they are stored as separate files that are accessed via a reference stored in the database. This category of errors can occur due to transient causes such as low memory and unrecoverable causes such as source blob files being deleted. Separate blob files containing IndexedDB data can end up being deleted because they show up as opaque files to users when they are using disk space recovery programs.
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:51]: This is all a Chromium implementation detail, so doesn't really belong in MDN like this. Different browsers could (and do) approach storing large values very differently. 

This page should lead with the error types (below), explain the cause (maybe mentioning that implementations can store values in different ways) and how developers should reason about what to do (e.g. retry if recovery is possible, wipe the database, etc)

File: files/en-us/web/api/idbrequest/error/index.md (hunk 1 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:17]: Yup, that reads much better. I've used your wording, in the next commit.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 2 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:23]: Right, OK, that makes sense. So, for the ones that still belong on this page, I have marked them as asynchronous errors, and added a bit of text that comments on how they can be handled, based on your text above. Let me know if you think that approach makes sense.

I've also improved the code example snippet on the page so that the event handler logs the error name and message, rather than just the error object.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 3 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:27]: Duh, yeah, sorry. I should have clocked this. I wrote most of the original IndexedDB MDN docs but have not looked at them for years! TBH, they could really use an overhaul, but that is beyond the scope of this PR.

For now, I've removed the errors that shouldn't be listed here, gone through the spec to see what methods they _are_ thrown on, and ensured they are included in their ref pages' "Exceptions" sections.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 4 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:28]: As before, I've removed it from this page and made sure that it is covered on the relevant method/property pages.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 5 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:24]: As before, I've removed it from this page and made sure that it is covered on the relevant method/property pages.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 6 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:36]: As before, I've removed it from this page and made sure that it is covered on the relevant method/property pages.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 8 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:40]: As before, I've removed it from this page and made sure that it is covered on the relevant method/property pages.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 9 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:42]: As before, I've removed it from this section and made sure that it is covered on the relevant method/property pages.

In this case, `InvalidStateError` can be thrown when you try to access the actual `error` property before the request is complete. So I've mentioned that on this page.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 10 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:45]: I've updated it to:

> Thrown if the application runs out of disk quota. In some cases, browsers prompt the user for more space, and the error is thrown if they decline the request. In other cases, the browser uses heuristics to determine whether more space can be assigned.

Does that work for you?

File: files/en-us/web/api/idbrequest/error/index.md (hunk 11 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:47]: As with `AbortError`, I've indicated that this error is asynchronous, and removed the "thrown" wording.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 2 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:23]: I've also removed the "thrown" language in each case.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 12 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:64]: Yup, totally agree. I think I can just remove this section. This page is about async errors related to requests; there is no need to highlight that other errors occur in other places.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 13 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:51]: Entirely agree. I've rephased/restructured it as suggested, and indicated that the storage description is just an example, which is the way Chrome does it.

File: files/en-us/web/api/idbcursor/request/index.md [hunk 14]
```diff
@@ -14,6 +14,11 @@ The **`request`** read-only property of the {{domxref("IDBCursor")}} interface r
 
 An {{domxref("IDBRequest")}} object instance.
 
+### Exceptions
+
+- `InvalidStateError` {{domxref("DOMException")}}
```
Comment (inexorabletash) [inline files/en-us/web/api/idbcursor/request/index.md:19]: `request` never throws

(`key`/`primaryKey`/`value` do while the cursor is advancing since it's in between positions, but `request` has a stable value for the lifetime of the `IDBCursor` so never throws)

File: files/en-us/web/api/idbdatabase/createobjectstore/index.md [hunk 15]
```diff
@@ -54,20 +54,22 @@ A new {{domxref("IDBObjectStore")}}.
 This method may raise a {{domxref("DOMException")}} with a `name` of
 one of the following types:
 
+- `ConstraintError` {{domxref("DOMException")}}
+  - : Thrown if an object store with the given name (based on a case-sensitive comparison)
+    already exists in the connected database.
+- `InvalidAccessError` {{domxref("DOMException")}}
+  - : Thrown if `autoIncrement` is set to true and `keyPath` is
+    either an empty string or an array containing an empty string.
```
Comment (inexorabletash) [inline files/en-us/web/api/idbdatabase/createobjectstore/index.md:62]: Should be: "either an empty string or an array."

The array contents don't matter.

File: files/en-us/web/api/idbdatabase/createobjectstore/index.md [hunk 16]
```diff
@@ -54,20 +54,22 @@ A new {{domxref("IDBObjectStore")}}.
...
 - `InvalidStateError` {{domxref("DOMException")}}
   - : Thrown if the method was not called from a
     `versionchange` transaction callback.
+- `SyntaxError`
+  - : Thrown if the `keyPath` option contains an invalid key path.
 - `TransactionInactiveError` {{domxref("DOMException")}}
   - : Thrown if a request is made on a source database that does not exist
```
Comment (inexorabletash) [inline files/en-us/web/api/idbdatabase/createobjectstore/index.md:69]: Also if the upgrade transaction has completed, or is processing a request.

File: files/en-us/web/api/idbrequest/error/index.md [hunk 17]
```diff
@@ -10,32 +10,38 @@ browser-compat: api.IDBRequest.error
...
+
+These errors are asynchronous, meaning that they can't be handled via [`try...catch`](/en-US/docs/Web/JavaScript/Reference/Statements/try...catch). However, if an `IDBRequest` has an {{domxref("IDBRequest.error", "error")}} event handler assigned, you can still inspect such errors by querying the request's `error` property via the event object, for example [`event.target.error.name`](/en-US/docs/Web/API/DOMException/name) or [`event.target.error.message`](/en-US/docs/Web/API/DOMException/message).
 
 - `AbortError`
   - : If you abort the transaction, then all requests still in progress receive this error.
 - `ConstraintError`
-  - : If you insert data that doesn't conform to a constraint.
-    It's an exception type for creating stores and indexes.
-    You get this error, for example, if you try to add a new key that already exists in the record.
+  - : Received if you insert data that doesn't conform to a constraint when creating stores and indexes.
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:25]: "when creating stores and indexes" may be misleading; when you're calling `createObjectStore()` or `createIndex()` those are synchronous ops that throw. You could reword this as "... doesn't conform to a constraint when populating stores."

File: files/en-us/web/api/idbrequest/error/index.md [hunk 18]
```diff
@@ -10,32 +10,38 @@ browser-compat: api.IDBRequest.error
...
+    For example, you will get this error if you try to add a new key that already exists in the record.
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:26]: nit: "store" not "record"

alternately: "a new key that already exists for a record in the store"

(a record is a key/value pair in a store)

File: files/en-us/web/api/idbrequest/error/index.md [hunk 19]
```diff
@@ -10,32 +10,38 @@ browser-compat: api.IDBRequest.error
...
+- `DataError` or `UnknownError`
+  - : Received for transient read failure errors, including general disk IO errors. See "Large value read failure errors" below for more details.
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:28]: Should "Large value read..." be a hyperlink to the section?

File: files/en-us/web/api/idbrequest/error/index.md [hunk 20]
```diff
@@ -10,32 +10,38 @@ browser-compat: api.IDBRequest.error
...
+- `NotFoundError` or `NotReadableError`
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:29]: Looking at the spec, we don't issue "NotFoundError" in this case. 

https://github.com/w3c/IndexedDB/pull/430/files only added "NotReadableError" for this case.

But maybe I missed something?

File: files/en-us/web/api/idbrequest/error/index.md [hunk 21]
```diff
@@ -10,32 +10,38 @@ browser-compat: api.IDBRequest.error
...
+Different IndexedDB implementations store large values in different ways. For example, in Chrome, large IndexedDB values are not stored directly in the underlying database; instead, they are stored as separate files that are accessed via a reference stored in the database. It has been observed that these separate files can end up being deleted because they show up as opaque files to users when they are using disk space recovery programs.
+
+Possible corrective actions for such cases might include notifying the user, deleting the entry from the database, then attempting to re-fetch the data from the server.
 
-In addition to the error codes sent to the {{ domxref("IDBRequest") }} object,
-asynchronous operations can also raise exceptions. The list describes problems that
-could occur when the request is being executed, but you might also encounter other
-problems when the request is being made. For example, if the result is accessed
-while the request is not completed, the `InvalidStateError` exception is thrown.
+More recent browser versions have changed the error types and improved the error messages to help developers distinguish between transient and unrecoverable cases.
```
Comment (inexorabletash) [inline files/en-us/web/api/idbrequest/error/index.md:44]: Maybe add: "Specifically, "NotReadableError" signals that the record is present in the database, but that the value could not be retrieved."

**Review by inexorabletash — COMMENTED** [review]
Looking good, a few more nits/suggestions.

File: files/en-us/web/api/idbdatabase/createobjectstore/index.md (hunk 16 above)
Comment (wbamberg) [reply files/en-us/web/api/idbdatabase/createobjectstore/index.md:69]: We should also remove that old Firefox bug info at this point.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 21 above)
Comment (wbamberg) [reply files/en-us/web/api/idbrequest/error/index.md:44]: I might have some more questions about this, but for now:

First, is this behavior cross-browser/per spec? If not, what do other browsers do?
 
Second, reading this, I'm not that sure how strong the link is between transient/unrecoverable and specifically "large file read" errors. So I see docs that say: `DataError` or `UnknownError` are for transient errors, and `NotFoundError` or `NotReadableError` are for unrecoverable errors. But these ask me to read the section on "Large value read failure errors" for more detail. But really that section seems to be about a specific case that arises from a specific browser's implementation in some situations. Can we have transient/unrecoverable errors  in other situations (presumably so, right?). If so, then this seems to overstate the connection.

Maybe it might be better to call this section "Transient and unrecoverable read errors", and say: sometimes read errors have transient causes (such as low memory) and sometimes they are unrecoverable (for example, if the data itself has been deleted). Developers will want to handle these cases differently. In the first case the error will be `DataError` or `UnknownError` (which one depends on ? browser version? browser engine?). In the second case, the error will be `NotFoundError` or `NotReadableError` (again, depending on ?).

Overall, I think it's best to approach this from the point of view of: what useful advice can we give a developer trying to write cross-browser code that can properly distinguish and handle transient and unrecoverable errors?

Comment (inexorabletash) [reply files/en-us/web/api/idbrequest/error/index.md:44]: This is where I might need to tag in others like @evanstade and @abhishek-shanthkumar and @asutherland

I believe recent Chromium changes were made to align behavior with Gecko (Firefox). But per https://github.com/w3c/IndexedDB/issues/423 there was a little bit of churn on all sides as we went from unspecified behavior to specified. Can the other folks I've tagged comment on what comprehensive documentation would be, and how much of a link there is between transient/unrecoverable vs. "large file" ?

Re:  _"it might be better to call this section "Transient and unrecoverable read errors""_ - agreed.

File: files/en-us/web/api/idbcursor/request/index.md (hunk 14 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbcursor/request/index.md:19]: Yup, the spec makes this pretty clear. I've removed this "Exceptions" section.

File: files/en-us/web/api/idbdatabase/createobjectstore/index.md (hunk 15 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbdatabase/createobjectstore/index.md:62]: updated

File: files/en-us/web/api/idbdatabase/createobjectstore/index.md (hunk 16 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbdatabase/createobjectstore/index.md:69]: I've added the missing information and removed the out-of-date Firefox info.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 17 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:25]: Sounds good; I've used your wording.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 18 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:26]: Makes sense; I've updated it to "...new key that already exists in the store."

File: files/en-us/web/api/idbrequest/error/index.md (hunk 20 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:29]: https://chromestatus.com/feature/5140210640486400 mentions `NotFoundError` being thrown in Chrome 130/131 for unrecoverable failures, but `NotReadableError` in Chrome 132, but I think I should just mention `NotReadableError` here regardless. Any historic Chrome info can go in the browser compat table.

Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:29]: Looking at the spec, `NotFoundError` is thrown on methods like `deleteObjectStore()` or `deleteIndex()` when the object store or index is not found. I've made sure it is mentioned in the exceptions lists on the appropriate pages.

Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:29]: I've also done the same thing with the transient error names — just left `UnknownError` in this list and removed `DataError`, making sure that instead, `DataError` is mentioned on the method pages that can throw it.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 21 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:44]: OK, I have:

- Added "Specifically, this error signals that the record is present in the database, but the value could not be retrieved." to the description of `NotReadableError` earlier on, as I think it makes more sense there.
- Updated the title to "Transient and unrecoverable read errors".
- Rearranged the structure a bit to be something more like Will suggested.
- Linked to the browser compat table for more information on browser-specific error types and behavior.

I'm hoping that we can get some more information about the state of things in other browsers to put here. Do you think this PR needs to be blocked on that info, or can we add it over time?

**Review by inexorabletash — COMMENTED** [review]
LGTM, one minor suggestion. I don't think we need to wait on perfect compat data before landing.

File: files/en-us/web/api/idbrequest/error/index.md [hunk 22]
```diff
@@ -14,28 +14,45 @@ request.
...
 - `QuotaExceededError`
-  - : If you run out of disk quota and the user declined to grant you more space.
+  - : Received if the application runs out of disk quota. In some cases, browsers prompt the user for more space, and the error is received if they decline the request. In other cases, the browser uses heuristics to determine whether more space can be assigned.
 - `UnknownError`
-  - : If the operation failed for reasons unrelated to the database itself.
-    A failure due to disk IO errors is such an example.
+  - : Received for transient read failure errors, including general disk IO errors. See [Transient and unrecoverable read errors](#transient_and_unrecoverable_read_errors) below for more details.
 - `VersionError`
-  - : If you try to open a database with a version lower than the one it already has.
+  - : If you try to open a database with a version lower than the one it already has, this error is received.
```
[the remark this answers is not in the document]
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md]: Good call; done.

File: files/en-us/web/api/idbrequest/error/index.md [hunk 23]
```diff
@@ -14,28 +14,45 @@ request.
...
-  - : If the operation failed for reasons unrelated to the database itself.
-    A failure due to disk IO errors is such an example.
+  - : Received for transient read failure errors, including general disk IO errors. See [Transient and unrecoverable read errors](#transient_and_unrecoverable_read_errors) below for more details.
 - `VersionError`
-  - : If you try to open a database with a version lower than the one it already has.
+  - : Received if you try to open a database with a version lower than the one it already has.
+
+### Transient and unrecoverable read errors
+
+Read errors occur when an IndexedDB stores large blob values (for example, audio files for an offline podcast app), and then subsequently fails to read those values even though the associated records are still in the database.
```
Comment (wbamberg) [inline files/en-us/web/api/idbrequest/error/index.md:37]: "Read errors occur when an IndexedDB stores large blob values" - is it true that read errors only occur for "large blob values"? Is that even really relevant here?

I think the key message is: some read errors are transient and some are unrecoverable. As a developer you probably need to know the difference, so you can handle them seprately. Transient errors are signalled with `UnknownError`, and unrecoverable errors are signalled with `NotReadableError`.

I think everything else here is not that important. You might want to refer to large values in passing, as an example of an unrecoverable error: "for example, sometimes the browser might store large values in a separate file, referenced from the database, and the referenced file might have been deleted by a disk space recovery program". You might also want to explain a bit about what a developer should do in each case, which could be (for the "unrecoverable errors" case) a version of your paragraph beginning "Possible corrective actions". If you do that though it would also be good to say what a developer should do in the other case, if there is anything to say here.

File: files/en-us/web/api/idbrequest/error/index.md [hunk 24]
```diff
@@ -14,28 +14,45 @@ request.
...
+
+This category of errors can occur due to:
+
+- Transient causes such as low memory, represented in the spec by the `UnknownError` type.
+- Unrecoverable causes such as source blob files being deleted, represented in the spec by the `NotReadableError` type.
+
+> [!NOTE]
+> The errors that represent these cases may vary across browsers. Check the [browser compatibility information](#browser_compatibility) for more details.
```
Comment (wbamberg) [inline files/en-us/web/api/idbrequest/error/index.md:45]: Not sure this note is needed. Everything might vary across browsers, you always need to check BCD.

File: files/en-us/web/api/idbrequest/error/index.md [hunk 25]
```diff
@@ -14,28 +14,45 @@ request.
...
+
+Such errors can manifest in different ways depending on browser, operating system, etc. For example, different IndexedDB implementations store large values in different ways. In Chrome, large IndexedDB values are not stored directly in the underlying database; instead, they are stored as separate files that are accessed via a reference stored in the database. It has been observed that these separate files can end up being deleted because they show up as opaque files to users when they are using disk space recovery programs, resulting in unrecoverable read errors when the IndexedDB is next accessed.
```
Comment (wbamberg) [inline files/en-us/web/api/idbrequest/error/index.md:47]: I don't think this paragraph is needed. It's also quite unclear "such errors" - errors such as what? Just read errors? There's no indication that this means specifically unrecoverable read errors. As I said above, I think a compressed version of this might be given as an example of "unrecoverable errors". I don't think the stuff about Chrome's implementation is needed or really in scope.

File: files/en-us/web/api/idbrequest/error/index.md [hunk 26]
```diff
@@ -14,28 +14,45 @@ request.
...
+
+Possible corrective actions for such cases might include notifying the user, deleting the entry from the database, then attempting to re-fetch the data from the server.
```
Comment (wbamberg) [inline files/en-us/web/api/idbrequest/error/index.md:49]: Again this is referring only to unrecoverable errors, but that's not clear from the context ("such cases" - which cases?)

**Review by wbamberg — CHANGES_REQUESTED** [review]

File: files/en-us/web/api/idbrequest/error/index.md [hunk 27]
```diff
@@ -14,28 +14,45 @@ request.
...
+  - : Received if you try to open a database with a version lower than the one it already has.
+
+### Transient and unrecoverable read errors
+
+Read errors occur when an IndexedDB stores large blob values (for example, audio files for an offline podcast app), and then subsequently fails to read those values even though the associated records are still in the database.
+
+This category of errors can occur due to:
+
+- Transient causes such as low memory, represented in the spec by the `UnknownError` type.
+- Unrecoverable causes such as source blob files being deleted, represented in the spec by the `NotReadableError` type.
```
Comment (wbamberg) [inline files/en-us/web/api/idbrequest/error/index.md:42]: FWIW I think "represented in the spec by" is a slightly strange way to word this. We always describe the spec. Why can't we say something like  "Transient causes such as low memory are signalled using the `UnknownError` type."

Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:42]: Yeah, you're right. I've updated these sentences to use "signalled by":

> - Transient causes such as low memory, signalled by an `UnknownError` type.
> - Unrecoverable causes such as source blob files being deleted, signalled by a `NotReadableError` type.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 26 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:49]: OK; updated to "Possible corrective actions for unrecoverable read errors might include..."

File: files/en-us/web/api/idbrequest/error/index.md (hunk 24 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:45]: OK, I've removed the note.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 25 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:47]: I've updated the paragraph to say "unrecoverable read errors" rather than "such errors" and removed the mention of Chrome, making the description more generic. However, I think it's useful to have an example of how an unrecoverable read error might occur, as readers might rightly think "well, how on earth could that happen?"

File: files/en-us/web/api/idbrequest/error/index.md (hunk 23 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md:37]: OK, so it is possible with any value, although it is more likely to happen with large values, for the kinds of reasons detailed in the example. So I've updated this sentence to "Read errors occur when an IndexedDB stores values and then subsequently fails to read those values..." and moved the example large file down to the example paragraph.

I've also restructured the section a bit, and attempted to fill in some more information on possible causes and solutions around transient errors. Saying that, it was hard to find much information around these.

@inexorabletash, can you check what I've written here and let me know if you think it makes sense? Thanks.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 27 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md]: Update - this ended up getting restructured significantly in a later edit.

**Review by wbamberg — APPROVED** [review]
👍 thank you Chris, I'm happy with this.

File: files/en-us/web/api/idbrequest/error/index.md (hunk 23 above)
Comment (chrisdavidmills) [reply files/en-us/web/api/idbrequest/error/index.md]: Update — I'm going to go ahead and merge this, so we can have it squared away before the end of the year. If you do have issues with what I have written, I'll fix it up in a new PR.
